File: dart/bogosort.dart
import 'dart:math';

void main() {
    BogoSort(GenerateRandomList(8), 100);
}

void BogoSort(List list, int printIterationsEvery) {
    int iterations = 0;
    PrintIteration(list, iterations);
    while (!IsSorted(list)) {
        iterations++;
        list = Shuffle(list);
        if (iterations % printIterationsEvery == 0) {
            PrintIteration(list, iterations);
        }
    }
    print("Result:");
    PrintIteration(list, iterations);
}

List Shuffle(List list) {
    Random rng = Random();
    for (int i = 0; i < list.length; i++)
    {
        int j = rng.nextInt(i + 1);
        int temp = list[i];
        list[i] = list[j];
        list[j] = temp;
    }
    return list;
}

bool IsSorted(List list) {
    for (int i = 1; i < list.length; i++) {
        if (list[i - 1] > list[i]) {
            return false;
        }
    }
    return true;
}

List GenerateRandomList(length) {
    Random rng = Random();
    List list = List<int>.filled(length, 0);
    for (int i = 0; i < length; i++) {
        list[i] = rng.nextInt(200) - 100;
    }
    return list;
}

void PrintIteration(List list, int iteration) {
    print("$iteration, $list");
}

File: rust/bogosort.rs
fn main() {
    bogo_sort(8, 100);
}

fn bogo_sort(length: i32, print_iterations_every: i32) {
    let mut iteration: i32 = 0;
    let mut list: Vec<i32> = generate_random_list(length);
    print_iteration(&list, iteration);
    while !is_sorted(&list, length) {
        iteration += 1;
        shuffle(&mut list, length);
        if iteration % print_iterations_every == 0 {
            print_iteration(&list, iteration);
        }
    }
    println!("Result:");
    print_iteration(&list, iteration);
}

fn shuffle(list: &mut Vec<i32>, length: i32) {
    for n in 0..length {
        let r: i32 = (rand::random::<i32>() % length).abs();
        let temp: i32 = list[n as usize];
        list[n as usize] = list[r as usize];
        list[r as usize] = temp;
    }
}

fn is_sorted(list: &Vec<i32>, length: i32) -> bool {
    for n in 1..length {
        if list[(n - 1) as usize] > list[n as usize] {
            return false;
        }
    }
    true
}

fn generate_random_list(length: i32) -> Vec<i32> {
    let mut list: Vec<i32> = Vec::new();
    for _n in 0..length {
        let x: i32 = rand::random();
        list.push(x % 100);
    }
    list
}

fn print_iteration(list: &Vec<i32>, iteration: i32) {
    println!("{:?}, {:?}", iteration, list);
}